### ISSUE TITLE
Linker detecting on Windows with ninja

### ISSUE THREAD

**Issue by ur4t** [mesonbuild/meson#6633]
**Describe the bug**

With clang (not clang-cl) targeting msvc, meson generates `build.ninja` file using `libfoo.dll.a` (mingw style) to link, while actually generated import library file is `libfoo.lib` (along with `libfoo.dll`).

**To Reproduce**

`meson.build` 
```meson
project('demo', 'cpp')
executable('demo', 'demo.cpp', link_with: shared_library('foo', 'foo.cpp'))
```
`foo.h` 
```cpp
#include <cstdio>

#ifdef __EXPORTS
#define DllFunc __declspec(dllexport)
#else
#define DllFunc __declspec(dllimport)
#endif

extern "C" DllFunc void foo();
```
`foo.cpp`
```cpp
#define __EXPORTS
#include "foo.h"
void foo() { puts("Test!"); }
```
`demo.cpp`
```cpp
#include "foo.h"

int main() { foo(); }
```
Sections of generated `build.ninja`
```ninja
rule cpp_LINKER
 command = clang++ @$out.rsp
 rspfile = $out.rsp
 rspfile_content = $ARGS -Wl,/MACHINE:X64 -Wl,/OUT:$out $in $LINK_ARGS
 description = Linking target $out.

build demo.exe: cpp_LINKER demo@exe/demo.cpp.obj | foo@sha/libfoo.dll.symbols
 LINK_ARGS = "-Wl,/nologo" "-Wl,/release" "-Wl,/nologo"
 "-Wl,/DEBUG" "-Wl,/PDB:demo.pdb" "libfoo.dll.a" "-mconsole"
```

**Expected behavior**

Sections of working `build.ninja`
```ninja
rule cpp_LINKER
 command = clang++ @$out.rsp
 rspfile = $out.rsp
 rspfile_content = $ARGS -Wl,/MACHINE:X64 -Wl,/OUT:$out $in $LINK_ARGS
 description = Linking target $out.

build demo.exe: cpp_LINKER demo@exe/demo.cpp.obj | foo@sha/libfoo.dll.symbols
 LINK_ARGS = "-Wl,/nologo" "-Wl,/release" "-Wl,/nologo"
 "-Wl,/DEBUG" "-Wl,/PDB:demo.pdb" "libfoo.lib" "-mconsole"
```

**System parameters**
* Clang 9.0.0 Target: x86_64-pc-windows-msvc
* MSVC 19.24
* Windows 10
* Python 3.8.1
* meson 0.53.1
* ninja 1.10.0

**Comment by ur4t** [comment]
Meson correctly reports ms link and gnu ld.bfd when initializing in corresponding environments. But actually both clang-msvc and clang-mingw report the linker is 'clang' in the logic to determine files' name. That is using the compiler's name, not the linker's.

**Comment by ur4t** [comment]
And rust on windows is also buggy. I'm trying to fix.

**Comment by ur4t** [comment]
Fixed with #6634.
